### PR TITLE
 [NEURON][TF][BUILD] - Release TF 2.10 with neuronx SDK 2.9.0

### PR DESCRIPTION
**Pull request by catalin-manciu-aws**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run

**NOTE: By default, docker builds are disabled. In order to build your container, please update dlc_developer_config.toml and specify the framework to build in "build_frameworks"**
- [ ] I have run builds/tests on commit <INSERT COMMIT ID> for my changes.

**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/dlc_tests/eks/eks_manifest_templates/pytorch/inference/single_node_neuronx_inference.yaml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+          limits:
+            cpu: "4"
+            memory: 4Gi
+          requests:
+            cpu: "1"
+            memory: 1Gi
+
+        resources:
+          limits:
+            aws.amazon.com/neuron: 1  # desired number of Inferentia devices.
```
Comment (aws-ceenugal) [inline test/dlc_tests/eks/eks_manifest_templates/pytorch/inference/single_node_neuronx_inference.yaml:55]: Inferentia => Trainium?

File: test/dlc_tests/eks/eks_manifest_templates/pytorch/inference/single_node_neuronx_inference.yaml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+          requests:
+            aws.amazon.com/neuron: 1  # desired number of Inferentia devices.
```
Comment (aws-ceenugal) [inline test/dlc_tests/eks/eks_manifest_templates/pytorch/inference/single_node_neuronx_inference.yaml:57]: Inferentia => Trainium?

File: test/dlc_tests/conftest.py [hunk 3]
```diff
@@ -409,7 +410,8 @@ def is_neuron_image(fixtures):
     :param request.fixturenames: active fixtures in the request
     :return: bool
     """
-    neuron_fixtures = ["tensorflow_inference_neuron", "mxnet_inference_neuron", "pytorch_inference_neuron"]
+    neuron_fixtures = ["tensorflow_inference_neuron", "tensorflow_inference_neuronx",
```
Comment (aws-ceenugal) [inline test/dlc_tests/conftest.py:413]: This conflicts with neuron_fixtures in this PR : https://github.com/aws/deep-learning-containers/pull/2776.
I am sure you are aware of it.

File: tensorflow/inference/buildspec-2-10-neuronx.yml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        target: neuron-monitor.sh
+      neuron-entrypoint:
+        source: docker/build_artifacts/neuron-entrypoint.py
+        target: neuron-entrypoint.py
+      deep_learning_container:
+        source: ../../src/deep_learning_container.py
+        target: deep_learning_container.py
+
+  images:
+    BuildNeuronTFInferencePy3DockerImage:
```
Comment (tejaschumbalkar) [inline tensorflow/inference/buildspec-2-10-neuronx.yml:35]: nit: rename to BuildNeuronX

File: tensorflow/inference/buildspec-2-10.yml [hunk 5]
```diff
@@ -3,6 +3,7 @@ region: &REGION <set-$REGION-in-environment>
 framework: &FRAMEWORK tensorflow
 version: &VERSION 2.10.1
 short_version: &SHORT_VERSION "2.10"
+device_type: &DEVICE_TYPE neuronx
```
Comment (tejaschumbalkar) [inline tensorflow/inference/buildspec-2-10.yml:6]: revert this change, this file does not concern with `neuron`

File: tensorflow/inference/buildspec-neuronx.yml [hunk 6]
```diff
@@ -0,0 +1,47 @@
+  account_id: &ACCOUNT_ID <set-$ACCOUNT_ID-in-environment>
+  region: &REGION <set-$REGION-in-environment>
+  framework: &FRAMEWORK tensorflow
+  short_version: &SHORT_VERSION "2.10"
+  version: &VERSION 2.10.1
+  device_type: &DEVICE_TYPE neuronx
```
Comment (tejaschumbalkar) [inline tensorflow/inference/buildspec-neuronx.yml:6]: nit: rename to BuildNeuronX

File: tensorflow/inference/buildspec-neuronx.yml [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+        target: neuron-monitor.sh
+      neuron-entrypoint:
+        source: docker/build_artifacts/neuron-entrypoint.py
+        target: neuron-entrypoint.py
+      deep_learning_container:
+        source: ../../src/deep_learning_container.py
+        target: deep_learning_container.py
+
+  images:
+    BuildNeuronTFInferencePy3DockerImage:
```
Comment (tejaschumbalkar) [inline tensorflow/inference/buildspec-neuronx.yml:35]: nit: rename to BuildNeuronX

File: tensorflow/inference/docker/2.10/py3/sdk2.9.0/Dockerfile.neuronx [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    aws-neuronx-tools=${NEURONX_TOOLS_VERSION} \
+    aws-neuronx-runtime-lib=${NEURONX_RUNTIME_LIB_VERSION} \
+ && apt-get clean \
+ && rm -rf /var/lib/apt/lists/*
+
+RUN ${PIP} --no-cache-dir install --upgrade \
+    pip \
+    setuptools
+
+# cython, falcon, gunicorn, grpc
```
Comment (tejaschumbalkar) [inline tensorflow/inference/docker/2.10/py3/sdk2.9.0/Dockerfile.neuronx:96]: Applies the comment here same as PT, unpin packages

File: test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py [hunk 9]
```diff
@@ -30,6 +31,13 @@
 def test_ec2_tensorflow_inference_neuron(tensorflow_inference_neuron, ec2_connection, region):
     run_ec2_tensorflow_inference(tensorflow_inference_neuron, ec2_connection, "8500", region)
 
+@pytest.mark.model("mnist")
+@pytest.mark.parametrize("ec2_instance_type", TF_EC2_NEURONX_ACCELERATOR_TYPE, indirect=True)
+# FIX ME: Sharing the AMI from neuron account to DLC account; use public DLAMI with inf1 support instead
+@pytest.mark.parametrize("ec2_instance_ami", [test_utils.UL20_PT_NEURON_US_WEST_2], indirect=True)
```
Comment (tejaschumbalkar) [inline test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py:37]: Lets fix the ami id if not already

File: test/dlc_tests/eks/eks_manifest_templates/pytorch/inference/single_node_neuronx_inference.yaml (hunk 1 above)
Comment (tejaschumbalkar) [reply test/dlc_tests/eks/eks_manifest_templates/pytorch/inference/single_node_neuronx_inference.yaml:55]: Please confirm that neuron devices are visible using `aws.amazon.com/neuron` for trn1 instances

File: test/dlc_tests/conftest.py (hunk 3 above)
Comment (catalin-manciu-aws) [reply test/dlc_tests/conftest.py:413]: Yes, I'll rebase one of the PRs once the other gets merged.

File: test/dlc_tests/eks/eks_manifest_templates/pytorch/inference/single_node_neuronx_inference.yaml (hunk 1 above)
Comment (catalin-manciu-aws) [reply test/dlc_tests/eks/eks_manifest_templates/pytorch/inference/single_node_neuronx_inference.yaml]: This is currently a placeholder since EKS tests are disabled for now

File: test/dlc_tests/eks/eks_manifest_templates/pytorch/inference/single_node_neuronx_inference.yaml (hunk 2 above)
Comment (catalin-manciu-aws) [reply test/dlc_tests/eks/eks_manifest_templates/pytorch/inference/single_node_neuronx_inference.yaml]: This is currently a placeholder since EKS tests are disabled for now, updated to "Neuron"

File: test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py (hunk 9 above)
Comment (catalin-manciu-aws) [reply test/dlc_tests/ec2/tensorflow/inference/test_tensorflow_inference.py:37]: It is fixed to a DLAMI that provides the correct neuron driver version (which is all we need), not to an exact ID (same as the others in that test, i didn't change anything)

File: tensorflow/inference/docker/2.10/py3/sdk2.9.0/Dockerfile.neuronx (hunk 8 above)
Comment (catalin-manciu-aws) [reply tensorflow/inference/docker/2.10/py3/sdk2.9.0/Dockerfile.neuronx:96]: Unpinned

File: tensorflow/inference/buildspec-neuronx.yml (hunk 7 above)
Comment (catalin-manciu-aws) [reply tensorflow/inference/buildspec-neuronx.yml]: Renamed

File: tensorflow/inference/buildspec-2-10.yml (hunk 5 above)
Comment (catalin-manciu-aws) [reply tensorflow/inference/buildspec-2-10.yml]: Reverted

File: tensorflow/inference/buildspec-2-10-neuronx.yml (hunk 4 above)
Comment (catalin-manciu-aws) [reply tensorflow/inference/buildspec-2-10-neuronx.yml]: Renamed

**Review by aws-ceenugal — APPROVED** [review]

**Review by tejaschumbalkar — APPROVED** [review]
